### ISSUE TITLE
Make the license script faster.

### ISSUE THREAD

**Issue by chinmaygarde** [flutter/flutter#118193]
From https://github.com/flutter/flutter/issues/117243#issuecomment-1358515664 by @Hixie 

```
The next thing I want to do is change the pattern-matching approach
entirely so as to not use RegExps, by first tokenizing the files and
then doing word-by-word pattern matching for the licenses. That should
improve performance by orders of magnitude.
```

**Comment by rmistry** [comment]
Also see related discussion in [skbug.com/14038](skbug.com/14038) about the license scripts now take ~1hr (2x longer) to run on the rollers into flutter/engine.

Pasting relevant comments from @Hixie here.
[#c14](https://bugs.chromium.org/p/skia/issues/detail?id=14038#c14):
```
The problem with the performance of the script is well-understood, it's a very expensive regular expression pattern. The long-term solution is to move away from RegExps and instead do manual tokenization. The premise is relatively simple, the execution is subtle and will take some time.

In the meantime, it is quite possible that we can make the expensive regular expression much cheaper; if you are interested in experimenting with that, the main thing I would start with is the `_lineBreak` pattern (which is substituted into many of the regular expressions to support arbitrary line breaks): https://github.com/flutter/engine/blob/main/tools/licenses/lib/patterns.dart#L140

The trick is finding a simpler pattern that still catches all the licenses.

Another possible approach to making things faster would be to look at the patterns and see if there are some that _start_ with the `_lineBreak` pattern, or have the `_lineBreak` pattern very near the start, and change those to not invoke `_lineBreak` until at least several words have been matched. This would like massively improve the backtracking that's going on today.
```

[#c16](https://bugs.chromium.org/p/skia/issues/detail?id=14038#c16):
```
For example we can probably improve the performance a lot by changing this (the most expensive pattern according to the log above):

  LicenseFileReferencePattern(
    firstPrefixIndex: 1,
    indentPrefixIndex: 2,
    fileIndex: 3,
    pattern: RegExp(
      kIndent +
      r"(?:This file is part of the Independent JPEG Group's software.)? "
      r'(?:It was modified by The libjpeg-turbo Project to include only code (?:and information)? relevant to libjpeg-turbo\.)? '
      r'For conditions of distribution and use, see the accompanying '
      r'(README.ijg)'.replaceAll(' ', _linebreakLoose),
      multiLine: true,
      caseSensitive: false,
    ),
  ),

...to something like:

  LicenseFileReferencePattern(
    firstPrefixIndex: 1,
    indentPrefixIndex: 2,
    fileIndex: 3,
    pattern: RegExp(
      kIndent +
      r"(?:This file is part of the Independent${_linebreakLoose}JPEG${_linebreakLoose}Group's${_linebreakLoose}software.)?${_linebreakLoose}"
      r'(?:It was modified by The libjpeg-turbo${_linebreakLoose}Project${_linebreakLoose}to${_linebreakLoose}include${_linebreakLoose}only${_linebreakLoose}code${_linebreakLoose}(?:and${_linebreakLoose}information)?${_linebreakLoose}relevant${_linebreakLoose}to${_linebreakLoose}libjpeg-turbo\.)?${_linebreakLoose}'
      r'For${_linebreakLoose}conditions${_linebreakLoose}of${_linebreakLoose}distribution${_linebreakLoose}and${_linebreakLoose}use,${_linebreakLoose}see${_linebreakLoose}the${_linebreakLoose}accompanying${_linebreakLoose}'
      r'(README.ijg)',
      multiLine: true,
      caseSensitive: false,
    ),
  ),

...i.e. removing the `_linebreakLoose` bits at the start of the two optional segments that start with the common words "This" and "It".
```

**Comment by rmistry** [comment]
@zanderso could we please increase the priority / severity of this issue (I do not have permission to change labels)?

Summarizing the specific issue for the rollers: All rollers deal with one checkout, when we have a regular roll going and other N manual rolls in the queue, they all need to run the license script sequentially. The 2x speed slowdown of the license script compounds and ends up slowing down all rolls (regular and manual). The license script prior to https://github.com/flutter/engine/pull/38148 used to take ~30 mins on the rollers. Now it takes ~1hr 10mins.

**Comment by zanderso** [comment]
I gave the suggestion from c16 a try, but it looks like it made things worse. I have not investigated the reason yet:

Before:
```
408829156μs tests -- /^((?:[-;@#<!.\\"/* ]*(?:REM[-;@#<!.\\"/* ]*)?[-;@#<!.\"/*]+)?)( *)(?:This *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*file *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*is *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*part *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*of *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*the *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*Independent *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*JPEG *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*Group's *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*software.)? *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*(?:It *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*was *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*modified *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*by *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*The *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*libjpeg-turbo *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*Project *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*to *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*include *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*only *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*code *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*(?:and *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*information)? *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*relevant *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*to *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*libjpeg-turbo\.)? *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*For *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*conditions *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*of *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*distribution *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*and *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*use, *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*see *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*the *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*accompanying *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*(README.ijg)/ (24197 tests, 98 matches, csReferencesByFilename (package:licenses/patterns.dart:411:14))
```

After:
```
414960319μs tests -- /^((?:[-;@#<!.\\"/* ]*(?:REM[-;@#<!.\\"/* ]*)?[-;@#<!.\"/*]+)?)( *)(?:This file is part of the *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*Independent *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*JPEG *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*Group's *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*software\.)? *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*(?:It was modified by The libjpeg-turbo *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*Project *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*to *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*include *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*only *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*code *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*(?:and *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*information)? *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*relevant *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*to *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*libjpeg-turbo\.)? *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*For conditions of distribution and use, *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*see *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*the *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*accompanying *(?:(?:\*/ *|[*#])?\r?\n(?:-|;|#|<|!|/|\*| |REM)*)*(README.ijg)/ (24197 tests, 98 matches, csReferencesByFilename (package:licenses/patterns.dart:411:14))
```

**Comment by zanderso** [comment]
Something something regular expressions something something now you have two problems =)

**Comment by mraleph** [comment]
See comments from @lrhn on https://github.com/dart-lang/sdk/issues/50301

tldr is that this code should be rewritten not to use regexps if you want to make it faster.
